packages/extension-base/src/services/balance-service/transfer/utils.ts

```diff
@@ -0,0 +1,118 @@
+// Copyright 2019-2022 @subwallet/extension-base
+// SPDX-License-Identifier: Apache-2.0
+
+import { ChainInfoMap } from '@subwallet/chain-list';
+import { ActionType, LEDGER_GENERIC_ALLOW_NETWORKS, ledgerMustCheckNetwork, ValidateRecipientParams, ValidationCondition } from '@subwallet/extension-base/services/balance-service/transfer/types';
+import { _isChainEvmCompatible, _isChainSubstrateCompatible, _isChainTonCompatible } from '@subwallet/extension-base/services/chain-service/utils';
+import { detectTranslate, isAddressAndChainCompatible, isSameAddress, reformatAddress } from '@subwallet/extension-base/utils';
+import { isAddress, isTonAddress } from '@subwallet/keyring';
+
+import { isEthereumAddress } from '@polkadot/util-crypto';
+
+function getConditions (validateRecipientParams: ValidateRecipientParams): ValidationCondition[] {
+  const { account, actionType, destChain, srcChain, toAddress } = validateRecipientParams;
+  const conditions: ValidationCondition[] = [];
+  const isSendAction = [ActionType.SEND_FUND, ActionType.SEND_NFT].includes(actionType);
+
+  conditions.push(ValidationCondition.IS_NOT_NULL);
+  conditions.push(ValidationCondition.IS_ADDRESS);
+  conditions.push(ValidationCondition.IS_VALID_ADDRESS);
+
+  if (!isEthereumAddress(toAddress) && !isTonAddress(toAddress)) {
+    // todo: need isSubstrateAddress util function to check exactly
+    // todo: add check advanced detection button
+    conditions.push(ValidationCondition.IS_VALID_SUBSTRATE_ADDRESS_FORMAT);
+  }
+
+  if (srcChain === destChain && isSendAction && !ChainInfoMap[destChain].tonInfo) {
+    conditions.push(ValidationCondition.IS_NOT_DUPLICATE_ADDRESS);
+  }
+
+  if (account?.isHardware) {
+    conditions.push(ValidationCondition.IS_SUPPORT_LEDGER_ACCOUNT);
+  }
+
+  return conditions;
+}
+
+function getValidation (conditions: ValidationCondition[], validateRecipientParams: ValidateRecipientParams): Promise<void> {
+  const { account, destChain, fromAddress, toAddress } = validateRecipientParams;
+  const destChainInfo = ChainInfoMap[destChain];
+
+  for (const condition of conditions) {
+    switch (condition) {
+      case ValidationCondition.IS_NOT_NULL: {
+        if (!toAddress) {
+          return Promise.reject(detectTranslate('Recipient address is required'));
```

Comment:
This function should only return the corresponding function to validate condition